src/python/pants/engine/build_files.py

```diff
@@ -201,90 +203,93 @@ def _hydrate(item_type, spec_path, **kwargs):
+
+    try:
+      all_addr_tgt_pairs = spec.all_address_target_pairs(addr_families_for_spec)
+    except Spec.AddressResolutionError as e:
+      raise AddressLookupError(e)
+    except SingleAddress.SingleAddressResolutionError as e:
+      _raise_did_you_mean(e.single_address_family, e.name)
+
+    all_matching_addresses = [
+      addr for (addr, tgt) in all_addr_tgt_pairs
+      if self._matches_target_address_pair(addr, tgt)
+    ]
+    # NB: This may be empty, as the result of filtering by tag and exclude patterns!
+    return all_matching_addresses
+
+  def all_matching_addresses(self):
+    """Return all addresses the given specs resolve to, within the given address families.
+
+    :raises: :class:`ResolveError` if:
      - there were no matching AddressFamilies, or
      - the Spec matches no addresses for SingleAddresses.
-  """
+    :raises: :class:`AddressLookupError` if no targets are matched for non-SingleAddress specs.
+    """
+    matched_addresses = []
+    for spec in self.specs.dependencies:
+      matched_addresses.extend(self._matching_addresses_for_spec(spec))
+    return matched_addresses
+
+
+@rule(MappedSpecs, [Select(AddressMapper), Select(Specs)])
+def map_specs(address_mapper, specs):
   # Capture a Snapshot covering all paths for these Specs, then group by directory.
   snapshot = yield Get(Snapshot, PathGlobs, _spec_to_globs(address_mapper, specs))
   dirnames = {dirname(f.stat.path) for f in snapshot.files}
   address_families = yield [Get(AddressFamily, Dir(d)) for d in dirnames]
 
-  # NB: `@memoized` does not work on local functions.
-  def by_directory():
-    if by_directory.cached is None:
-      by_directory.cached = {af.namespace: af for af in address_families}
-    return by_directory.cached
-  by_directory.cached = None
-
-  def raise_empty_address_family(spec):
-    raise ResolveError('Path "{}" does not contain any BUILD files.'.format(spec.directory))
-
-  def exclude_address(spec):
-    if specs.exclude_patterns:
-      return any(p.search(spec) is not None for p in specs.exclude_patterns_memo())
-    return False
-
-  def filter_for_tag(tag):
-    return lambda t: tag in [str(t_tag) for t_tag in t.kwargs().get("tags", [])]
-
-  include_target = wrap_filters(create_filters(specs.tags if specs.tags else '', filter_for_tag))
-
-  addresses = []
-  addresses_to_exclude = []
-  included = set()
-  def include(address_families, predicate=None):
-    matched = False
-    for af in address_families:
-      for (a, t) in af.addressables.items():
-        if (predicate is None or predicate(a)):
-          should_exclude_address = exclude_address(a.spec)
-          if should_exclude_address:
-            addresses_to_exclude.append(a)
-          if include_target(t) and (not should_exclude_address):
-            matched = True
-            if a not in included:
-              addresses.append(a)
-              included.add(a)
-    return matched
+  yield MappedSpecs(address_families, specs)
```

Comment:
While splitting this rule into two makes testing easier, it's basically pure overhead at runtime.

The method is called ~once per run though, so not a big deal. But if it's possible to remove the testing boilerplate by just adding a test helper method instead, that would be preferable.

Reply:
The rules were merged back together in a694488, and a helper method was created in e47a710!